en/05_Contributing/05_Coding_Conventions/01_PHP_Coding_Conventions.md

```diff
@@ -136,6 +136,133 @@ class MyClass extends Class
+- Note that there is a lot of legacy Silverstripe code where protected was used over private.
+
+## Interfaces
+
+- If there are or there's a realistic expection that there will be multiple class implementations of something then create an interface and use that for parameter and return types, and as the service key for injector.
+- If there is only a single implemention and no forseeable expectation of more implementaitons being added, then do not create an interface for the single class implementation because this would only be adding unecessarily complexity.
+
+## Typing
+
+- Aim to strongly type all new properties, method parameters, and method return types.
+- The exception to this is adding a new method or property to a class which is entirely untyped and adding typed API would be very inconsistent with the rest of the file. In this instance use PHP docblocks to define types.
+- If there are duplicate strong types and docblock types then remove the old docblock types.
+- Do not strongly type existing API outside of a new major version as it's an API breakage.
+
+## Mixed types
+
+- Where possible avoid using mixed types, including type unions, because this leads to complicated and brittle code.
+- Try to avoid using nullable scalars and instead use the default value of the scalar instead of null. One exception to this is for when `null` means 'uninitialised`.
+- Avoid the use of the splat `...` parameter as this makes API enforcement difficult. If you need to support multiple arguments then use an array parameter type.
+
+## Object instantiation
+
+- Use Injector to instantiate objects rather than the `new` keyword. This is done as it allows projects to swap out classes with different implementations.
+- Note there is only a minimal performance penalty for this, though there is a downside that IDE type hinting stops working. During development you may wish to use the new keyword for this reason and change it to use injector before submitting your PR.
+- For classes that use the Injectable trait use the `::create()` static method.
+- For objects that do not have the Injectable trait, use `Injector::inst()->get({classname})`
+- This also includes instantiation of third party classes `<<< @todo why? discuss`
+- The exception to this is for unit-tests where you should use the `new` because there you need to ensure you're testing a specific class
+
+## Extensions and traits
+
+- Use the DataExtension class for extending DataObjects.
+- Use the Extension class for extending non-DataObjects.
+- Typically you shouldn't use native PHP traits. A common gotcha is mistakenly using a trait on a DataObject to add a private static config property. This doesn't work because private static config properties are a Silverstripe specific thing and the Silverstripe DataExtension will correctly merge them. However traits are unable to merge private static config properites.
```

Comment:
```suggestion
- Typically you shouldn't use native PHP traits. A common gotcha is mistakenly using a trait on a DataObject to add a private static config property. This doesn't work because private static config properties are a Silverstripe specific thing and the Silverstripe DataExtension will correctly merge them. However traits are unable to merge private static config properties.
```

Reply:
I don't necessarily agree with this standard. I'd prefer instead that we have clearer guidelines about when to use a trait for composable functionality.

e.g. the `Form::addExtraClass()` and related methods are also implemented on `FormField` - these would be great as a trait because we reduce code reuse, there's no expectation that these methods should be dynamically added to any third-party classes, they don't require adding/changing configuration properties or any knowledge of the _type_ of class they're being applied to, and critically they add a new property to the class. Anything that adds a per-instance property cannot be done in an `Extension` because extensions are singletons.

So maybe something like this instead:

```suggestion
- Use a trait instead of an `Extension` when:
    - the composable functionality relies on per-instance properties (these can't be added to extensions)
    - the composable functionality doesn't add or modify configuration properties or values
    - there's no expectation that the composable functionality should be applied to third-party code.
- Use an `Extension` instead of a trait when:
    - the composable functionality doesn't rely on per-instance properties
    - the composable functionality adds or modifies configuration properties or values
    - there's an expectation that the composable functionality should be applied to third-party code.
```

Reply:
Updated